### PR TITLE
Add support for the WebAssembly tail calls proposal in Wasmtime

### PR DESCRIPTION
**Pull request by fitzgen**
### [Rendered](https://github.com/fitzgen/rfcs-2/blob/tail-calls/accepted/tail-calls.md)

### CODE REVIEW

File: accepted/tail-calls.md
Comment (bjorn3) [inline accepted/tail-calls.md:489]: They can have the same name on all architectures, right? Also isn't it still necessary to have a windows and non-windows version to account for any different redzone sizes (0 on x86_64 windows, 128 bytes I believe on x86_64 Unix) and unwinder impls? Those are part of the platform abi you aren't allowed to violate. Using a too big redzone will cause interrupts to mess with the stack and using the wrong unwinder impl breaks backtraces.

**Comment by cfallin** [comment]
This is fantastic! Thank you for writing this up in such detail!

One thought that occurred to me as I read through the args-below-SP approach -- which I think overall is brilliant, to be clear, in that it allows us to have zero-cost cleanup -- is with respect to volatility below SP. At least on x86-64, the *system* ABIs define either stack memory below rsp-128 (SysV, given redzone) or stack memory below rsp (Windows Fastcall) as volatile. We're defining our own new ABI here, so it's fine to say that we define memory under RSP as under the control of the current function and clobbered only at calls. But it's worth verifying that we aren't creating any dangerous interactions with the system, since the reason for the redzone to exist is to protect against *async* clobbers (signal handlers and the like):

- On Linux and macOS, I believe we set an alternate signal stack, so if we were to take a signal while we're setting up stack args below RSP but before RSP is decremented in the callee, the signal handler's frame would not overwrite the args.
- On Windows, however, there's no redzone and everything is considered volatile to even async clobbering, as per [this post](https://eli.thegreenplace.net/2011/09/06/stack-frame-layout-on-x86-64) at least:

  > [On Windows] There is no concept of "red zone" whatsoever. In fact, the ABI explicitly states that the area beyond rsp is considered volatile and unsafe to use. The OS, debuggers or interrupt handlers may overwrite this area.

I wonder if there's a way to avoid this? "Interrupt handlers" here seems dubious in userspace (they should use the ring-0 stack) but I know that Windows also does async "upcalls" for IO completions and that sort of thing. Can we set this to happen on a different stack? @peterhuene any thoughts here?

**Comment by joshtriplett** [comment]
I've had a fair bit of experience with the red zone on various targets, and you can't in general force the target to respect the red zone if it natively doesn't. I think there's no way around this on Windows, or on other targets that use the windows calling convention, such as UEFI. So either we should avoid using the red zone on all targets, or keep using different ABIs on different targets so that we can be more efficient on Linux by using the red zone.

File: accepted/tail-calls.md
Comment (Amanieu) [inline accepted/tail-calls.md:1170]: This layout is actually better than the one proposed in this RFC because it avoids the issue at the call site where data is written below the stack pointer. Even though wasmtime itself uses an alternate signal stack, the embedder could also have set up another signal handler without `SA_ONSTACK` which would not use the alternate signal stack.

No copying of the return address is actually necessary since x86 is the only architecture which combines a pop of the return address in the same instruction as the return itself. Other architectures use a return address register which can be popped first, after which the stack arguments can be freed before actually returning.

File: accepted/tail-calls.md
Comment (programmerjake) [inline accepted/tail-calls.md:1024]:
```suggestion
The caller is a `wasmtime::Func::call` invocation and the callee is a
```

File: accepted/tail-calls.md
Comment (programmerjake) [inline accepted/tail-calls.md:1239]: an alternative idea to diy-ing parallel copies would be to generate a fake entry block so regalloc2 register allocates for all alternative entry points (trampolines fused to the function) simultaneously. this would allow regalloc2 to potentially make better register allocations, since it register allocates for all the prologues/trampolines simultaneously

something like (can be adapted to other calling conventions as necessary):
```
fake-entry: # entry block for reg-alloc
    fake-jump entry-wasm, #[cold] entry-native, #[cold] entry-array # idk if branch hints are supported
entry-wasm: # label used as function ptr for wasm calls
    push fp
    fp = sp
    alloc-stack
    args = input-regs-and-stack-slots
    jump entry(args)
entry-native: # label used as function ptr for native calls
    push-native-return-trampoline-address  # makes return insn return into trampoline for cleanup
    push fp
    fp = sp
    alloc-stack
    args = input-regs-and-stack-slots
    save-fp-sp-etc
    jump entry(args)
entry-array:
    push-array-return-trampoline-address  # makes return insn return into trampoline for cleanup
    push fp
    fp = sp
    alloc-stack
    args = load-from-array
    save-fp-sp-etc
    jump entry(args)
entry(args):
    ...normal-wasm-body...
return:
    sp = fp
    pop fp
    ret
```

push-*-trampoline-address could be compiled to:
x86_64
```
    call rest-of-entry # prime branch predictor
return-trampoline:
    cleanup...
    ret
```
everything else:
```
    save lr
    lr = call rest-of-entry # prime branch predictor
return-trampoline:
    cleanup...
    ret lr
```

Comment (bjorn3) [reply accepted/tail-calls.md:1239]: I have been thinking about something like this in the past too even before this rfc. Exception support requires multiple exit points which are observably different. It isn't much of a stretch to extend this to multiple entry points for a function as well even if the implementation may be different.

Comment (programmerjake) [reply accepted/tail-calls.md:1239]: i added the fake entry block because most dominator tree calculations require a root, so regalloc2 likely also requires that

Comment (Amanieu) [reply accepted/tail-calls.md:1239]: I've used regalloc2 with multiple entry points in my own code. The design I used is based on [EntrySwitch](https://webkit.org/docs/b3/intermediate-representation.html) from Webkit's B3 compiler:

> B3 supports multiple procedure entrypoints. The way you create multiple entrypoints is by placing an EntrySwitch at the end of the root block. The root block must then have a successor for each entrypoint. Additionally, you must tell the Procedure how many entrypoints you want.
>
> [...]
>
> EntrySwitch behaves as if each Procedure has a variable called Entry. Each physical entrypoint sets Entry to the index of that entrypoint (so 0, 1, or 2, above) and jumps to the root block. EntrySwitch is just a switch on the Entry variable. Transformations over code that uses EntrySwitch are valid so long as they don't change the procedure's behavior under these semantics.
>
> EntrySwitch is implemented without any actual variables or switching. Instead, all code that lies on some path from the root block to some EntrySwitch is cloned for each entrypoint. This lowering is done as a very late phase in Air, so most of the compiler does not have to know anything about entrypoints. Most of the compiler treats EntrySwitch as an opaque control flow construct. EntrySwitch lowering is allowed to clone an arbitrary amount of code. However, normal uses of EntrySwitch will place it at the end of an empty root block and B3 will only hoist a handful of things above EntrySwitch. This leads to only a small amount of cloned code in practice.

File: accepted/tail-calls.md
Comment (eqrion) [inline accepted/tail-calls.md:1141]: Our plans are still in flux here, we're currently prototyping out some other approaches other than what Lars prototyped in that document.

For extra context too, our primary complexity here is that our ABI has some global state (pinned registers and some JS::Activation state) that must be kept in sync with our current module instance. Upon calls that might cross module instances (imports/indirect/call_ref), the caller performs a context switch before and after the call to maintain this global state. We don't do this work on direct calls, as we statically whether it will be cross-instance. This biases us in favor of direct calls, as they're assumed to be the critical path.

With tail calls however, we can replace a same-instance callee with a cross-instance callee and when that cross-instance callee returns we need to do a context-switch, but the direct caller is not set up to do that.

Lars' scheme adds an extra frame on the first tail call to prevent this from happening. This has the noted complexity, so we're evaluating some other options too.

**Comment by fitzgen** [comment]
I must admit, as many of you have pointed out, that I did not fully think through the interaction with the OS and the red zone. I thought we would be in the clear because of our `sigaltstack` usage but that doesn't apply to Windows.

Given that, I think we will have to settle for this frame layout:

```
                                               FP                     SP
                                               |                      |
                                               V                      V
[ ...caller-frame... | ...stack-args... | RA | FP' | ...callee-frame... ]
```

This doesn't allow the callee to clean up stack args when resetting SP to FP, but that isn't too bad. Luckily, x86-64 has the `ret <imm16>` opcode mentioned in the RFC, and our other targets all use a return address register (IIUC) so we wouldn't need to copy the return address around on the stack and kill speculation performance and compatibility with hardware CFI features.

I will update the RFC in a bit.

File: accepted/tail-calls.md
Comment (fitzgen) [reply accepted/tail-calls.md:1141]: Good to know, thanks @eqrion. Interested to hear where y'all end up in the design space.

File: accepted/tail-calls.md
Comment (fitzgen) [reply accepted/tail-calls.md]: Thanks!

**Comment by fitzgen** [comment]
Okay, I've updated the proposed calling conventions to avoid writing stack arguments past `SP`.

I don't see a way  around copying the return address around on the stack in tail calls when the tail caller and tail callee have different numbers of stack arguments. If anyone has ideas, I'm all ears.

I've added an open question about the interaction this might have with hardware CFI features (specifically Intel CET and its shadow stack, I don't think this is incompatible with ARM's pointer authentication, AFAIK).

File: accepted/tail-calls.md
Comment (programmerjake) [inline accepted/tail-calls.md:768]: from what I know, most call stack branch predictors don't care if you copy the return address around, all they care is that the return branches to where they expect. This is particularly true on ISAs where the return address is stored in the Link Register (nearly every common ISA except x86-32/64), since compilers commonly move that around and move it back when returning.

File: accepted/tail-calls.md
Comment (programmerjake) [inline accepted/tail-calls.md:1249]: afaict from reading https://www.felixcloutier.com/x86/ret.html Intel CET doesn't care which RSP it reads from, all it cares about is if the loaded value matches the value from the shadow stack.

Comment (cfallin) [reply accepted/tail-calls.md:1249]: Yup, and this should be the case given that we implement the tail call itself with a jmp rather than call -- the only `call` the machine sees is the original call, and the return address will match that call. (In other words a call/tail_call/tail_call/.../ret sequence is a push-shadow/nop/nop/.../pop-shadow sequence on the shadow stack, so everything is still properly nested.)

File: accepted/tail-calls.md
Comment (cfallin) [reply accepted/tail-calls.md:768]: Indeed, in the CPU frontend the machine just sees `call` and `ret` at various RIPs, and doesn't know RSP yet; call pushes on the predictor's internal stack and ret pops (slight simplification wrt misspeculation cleanup but...); the load from the actual in-memory stack will happen in the backend and be verified against the popped (predicted) value but in principle could have been loaded from anywhere.

File: accepted/tail-calls.md
Comment (fitzgen) [reply accepted/tail-calls.md:1249]: Great, I'll remove this concern.

File: accepted/tail-calls.md
Comment (fitzgen) [reply accepted/tail-calls.md:768]: Will remove this concern as well.

**Comment by cfallin** [comment]
One aspect that might be useful to hash out in the RFC is the rough implementation strategy for the actual tail-calling operation, I think. This is pretty subtle in general, because it's a transfer of an arbitrary number of args (more than can fit in registers in the general case) from one's own state, including spillslots in one's own frame, into a sequence of stack args that overwrite that frame.

(The RFC starts to get at this when discussing stack-arg permutation and regalloc2's solution to the parallel-move problem, but it's actually an issue even when the caller has no stack-args. The proposed solution is to "allocate more space for temporaries than necessary" and I guess a tl;dr of this comment is a request to fill that out, in a way!)

It seems to me that in the general case, this requires building the new stack-arg list *below* the current frame then moving it upward (and putting the return address beneath it), unless someone else sees another way? Consider the case:

* `f` has no stack-args (its signature is `() -> i32` for example), but has 1000 live values in its body;
* `f` tail-calls to `g`, with signature `(i32, i32, ... /* 1000 i32s total */) -> i32`.

Currently the ABI code constructs the call frame by, in pre-regalloc code, decrementing SP, doing a bunch of stores into stack-arg slots, then the call with reg-args constrained appropriately. Those stores are safe to regalloc even if we have to reload their args from spillslots because the slots are above the arg-list we're constructing. A tail-call without the "move upward" scheme would not satisfy this condition though.

So it seems to me that we could introduce a mode to the caller-side ABI implementation where we build stack args as normal, but then just before the `call` instruction, do an inline `memmove` of data (stack args and return address) upward on top of our frame, clobbering it. All of this needs to be *one VCode instruction* because regalloc cannot be allowed to insert anything in the middle: its frame is no longer valid. We also should consider the intermediate state during this move, with respect to an FP-walk from a profiler async interrupt. I.e., the two properties that must be true at every point between instructions are (i) no valid stack data below SP, (ii) FP points to a valid frame-chain.

I think the following would work:

Generate pre-regalloc vcode instructions to:
- Decrement SP with size of stack-args for tail-callee;
- For each stack arg, store arg vreg (these might load values from spillslots!);
- Do a tailcall, with immediate parameters for existing stack-arg length and new stack-arg length, and register args to constrain reg-passed func arguments as we do today.

After steps 1 and 2, both conditions are true: no data below SP, and FP and data it points to is unchanged.

That tailcall inst then generates a sequence:
- Load return address from [FP+8] into a temporary. This needs to come first before we lose our old FP below.
- Load FP from [FP]. This lets us start to clobber the existing frame, and effectively (temporarily) removes this frame from a stack-walk if an interrupt were to happen at this point. (This also implies that we cannot have any safepoints/GC walks during this sequence, but that's to be expected I think.)
- Store return address into positive offset from SP: [SP + old frame length + old stack arg length - new stack arg length] (double-check me there, may be off by a word or two?). This needs to come after we load FP above, because if we're unlucky, it might overwrite the old FP. (TODO: this might be impossible because of 16-alignment of the (FP, PC) tuple, so one is always in odd word and other is even?)
- Do a memmove of args from SP offset 0 to SP offset N.
- Only now, increment SP to just below return address.
- Then jump!

Thoughts? Maybe there's a simpler way that's still safe at each instruction?

**Comment by cfallin** [comment]
And, to illustrate how subtle this is, I think the "store return address" needs to come *after* the memmove, not before, because otherwise it could also overwrite the args we're about to memmove.

The return address temporary needs to be an "early def" regalloc arg on the tail-call vcode inst, as well, I think.

**Comment by bnjbvr** [comment]
@fitzgen 

> Nothing requires that this be implemented inside Cranelift or in the backends or calling convention code.

Out of curiosity: are you saying it would be possible then to take this stack information gathered in userland and inject it into error reports and the debugging infra (for printing it in a backtrace obtained with `bt`, for instance)? I'm totally out of my league here, so really asking about feasibility.

> Wasmtime itself doesn't provide a profiler

Right, what I meant really was "generates profiling support for external tools like perf and vtune". Anyways, it seems we agree that it shouldn't change anything here.

> when at an actual source position

Right, that makes sense. Thanks for the confirmation!

**Comment by fitzgen** [comment]
@cfallin, I had initially been hoping that I could just get `regalloc2` to put things in the right places via constraints on the vregs, but it seems that stack slot constraints aren't mature enough for that right now. Given that, I think the build-the-stack-args-below-and-memmove-them-over approach that you describe is the right initial approach. I'll add some text to the RFC.

**Comment by fitzgen** [comment]
@bnjbvr 

> Out of curiosity: are you saying it would be possible then to take this stack information gathered in userland and inject it into error reports and the debugging infra (for printing it in a backtrace obtained with `bt`, for instance)? I'm totally out of my league here, so really asking about feasibility.

I don't think we can get `gdb`/`lldb` to integrate the shadow chicken-style log into their backtraces; at least I'm not aware of how we could. I guess DWARF CFI information can hold DWARF expressions which are a turing complete bytecode language, but... I wouldn't want to lose hair trying to hack this together and I bet it would be sloooooow.

I was talking about making captured backtraces in Wasmtime (e.g. `wasmtime::WasmBacktrace`) integrate the "lost" tail calls. This should be possible at the Wasmtime level, without any additional help from inside of Cranelift.

**Comment by bjorn3** [comment]
> I don't think we can get gdb/lldb to integrate the shadow chicken-style log into their backtraces

Apart from hacking something together with DWARF CFI information as you said (maybe Cranelift could get a DWARF CFI backend?), could write a plugin for gdb and lldb. Gdb supports frame filters which can arbitrarily modify the list of stack frames presented to the user: https://sourceware.org/gdb/onlinedocs/gdb/Writing-a-Frame-Filter.html#Writing-a-Frame-Filter As for lldb I'm not sure.

**Comment by fitzgen** [comment]
Ah of course, I forgot about `gdb`'s python plugins. That would probably be the best approach.

**Comment by uweigand** [comment]
Hi @fitzgen, here's my thoughts on the proposed new ABI.  First of all, thanks for the detailed write-up, this helps me understand in particular the situation with the various trampolines from and to native code.  I also agree that having some form of callee-pops-arguments rule is most likely necessary to implement tail calls with the required properties.  [ B.t.w. callee-pops seems fundamentally incompatible with variable argument lists, so I assume this ABI makes use of the fact that those are not supported in wasmtime, right? ]

That said, the specifics of the proposed ABI seem very prescriptive to me.  Read literally, it seems to imply that we basically take an ABI specification that makes sense on x86, and require all other architectures to implement it in exactly the same way, with the only "freedom of choice" being which registers are used for argument passing.  While I guess this might be possible, on many platforms this will lead to an implementation that is less efficient, and also more fundamentally incompatible with their "native" ABI, e.g. in a way that will break external tooling like debuggers or profilers.

---

Talking only about s390x now, here's some of the specific concerns I see:

### Return address
The instruction used to implement calls does not push the return address to the stack, but saves it into a register.  (That's actually the case on just about any non-x86 architecture I'm aware of.)   This means on the one hand that there's no concern about moving that adddress to a different slot during tail calls.

On the other hand, there's no fixed location defined by the ISA where to find return addresses on the stack, so if we want allow external tools to find them during unwinding, we need to instead follow rules defined by the platform ABI (see also the next points).

### Frame pointer
Related to (non-DWARF-CFI-based) stack unwinding, on our platform -like several others- there is no notion of a "frame pointer".  Or rather, while compilers do implement "frame pointers", their purpose is strictly to allow addressing local stack slots in the presence of dynamic stack reallocation (alloca), they are fundamentally not suitable for stack unwinding.  [Since there is no alloca in wasmtime, this means frame pointers are irrelevant for wasmtime. ]

The two main reasons why frame pointers, even if we were to force their use, are useless for unwinding are:

1. The ABI does not define *which* register is to be used as a frame pointer, and where a caller's version of that register needs to be saved; rather, this is up to each compiler to decide.
2. Even if we knew, the frame pointer points to the *bottom* of a stack frame, i.e. exactly where the stack pointer points before any alloca, and not to the top of the stack frame like on x86 - only the latter property allows it to be also used for unwinding.  This implementation choice was made for good reasons based on the ISA: with a frame pointer that points to the top of the frame, accesses to stack slots need to use negative displacements, and with our ISA, those are less efficient to implement than positive displacements.

We *do* have a mechanism that is defined by the ABI and implemented by external tools, which can be used for DWARF-less unwinding similarly to how the frame pointer chain can be used on x86.  That mechanism uses "stack backchains" which work as follows:

1. The initial stack backchain value can at all times be found in the 64-bit word addressed by the stack pointer.
2. Each stack backchain value points to the location on the stack containing the backchain of the next-higher stack frame.

This is currently implemented in wasmtime as well, and used for fast unwinding.  Note that this can be implemented very efficiently by simply storing a copy of the old stack pointer value at the new stack pointer location after the stack pointer was decremented in the function prologue.  The implementation is compatible with the system ABI so external tools are able to use it as well without any adaptation.

However, if we want to continue using that method in the presence of tail calls, the tail call ABI needs to allow for backchain values to continue to be placed at this specific location (at the very bottom of the stack).

Note that if we don't want to unwind only the stack frame addresses, but also the return address in each frame, those also need to be stored at well-known locations, see the next point.

### Register save area

In addition to the stack backchain word, the system ABI actually defines a whole area of size 160 bytes at the very bottom of each stack frame (the bottom-most word of which is reserved to hold the backchain).  This area must always be allocated by the caller, and can be freely used by the callee.  Its typical use is to save GPRs via a store-multiple instruction, which in this way can be done *before* the caller allocates its own stack frame.  This is both more efficient, and also provides well-known locations for at least the return address (see above) - the stack backchain points to the base address of the register save area, and the return address register is saved at a well-known offset within that area.

While the specific register save area mechanism does not necessarily have to be implemented in exactly the same way in our own variant ABI, if we want external unwinding tools to continue to work, at least the location of saved return addresses has to stay compatible.

### Allocating stack arguments

On Intel, stack arguments to a called function are typically pushed one by one to the stack.  This makes sense in the presence of a "push" instruction, and an ABI that was designed with use of that instruction in mind.   On our platform, we don't have a push instruction, and furthermore we do not want to continually manipulate the stack pointer - recall that the stack pointer is supposed to always point to the backchain word, so any change of the stack pointer normally needs to copy the backchain word to the new location.   Therefore, in the platform ABI this operation only takes place when absolutely necessary (i.e. to implement alloca), and in wasmtime it *never* takes place.

Instead, the function prologue already allocates an argument area at the bottom of the stack (i.e. above the register save area reserved for our callees, but below any local stack slots) that is big enough to hold all required arguments for any calls performed in this function (as determined statically during compilation of the function).   Stack arguments are then simply moved into their appropriate location before the call, without ever manipulating the stack pointer.  [ This method is used by many non-x86 architectures, even on Arm in other compilers. ]

While this is not necessarily *required* by the ABI and can be changed without breaking compatibility with external tools  -- and will *have* to change in some way in order to implement the required tail call properties -- we should still try to keep any necessary copying of the stack backchain word to a minimum.

---

For these reasons, I would argue for being less prescriptive in the ABI definition, and in fact leave most details to each architecture - only prescribe the core properties that are needed to implement all required goals.

That said, I'd still need to think a bit more what the best implementation on s390x would be even under those circumstances.  Maintaining the stack backchain does appear rather inefficient in a callee-pops world, but if we change that, we lose compatibility with external tools -- need to make a decision here what's more important.

One thing seems clear: it is probably unavoidable that regular calls, at least for functions that require stack arguments, will have to take some performance hit.  This is simply because there will be some need to adjust the stack pointer before the call, which is not done today.  (The same also applies to returns.  While those restore the old stack pointer already today, of course, this is done nearly for free as part of a load-multiple operation.  In the new scheme, we cannot just restore the old stack pointer but have to adjust it in addition, which will be at least one extra instruction.)   If we do maintain the backchain in a manner compatible with the ABI, each of those new stack adjust operations will *also* have to involve storing the backchain again.

A number of approaches to be considered:

1. As close as possible to the existing ABI: Caller's stack frame looks the same as today except no reserved space for arguments (but including the callee register save area + backchain).   Before the call, decrement the stack pointer to make room for arguments and re-write the backchain (as if doing an alloca).   During return, increment the stack pointer and re-write the backchain (as if freeing an alloca area).
2. Keep arguments *below* the register save area.  Same as above except not rewriting the backchain.  The callee needs to know where to find the incoming save area (no longer at SP, but at SP + incoming stack argument size).  Drawback: Initial backchain pointer is not maintained at SP during the call sequence itself, which might throw off asynchronous backtracing (e.g. system profilers).
3. Variant of the above where we don't maintain a register save area at all, but allocate it during the call sequence (either on the caller or callee side).  Could still keep the initial backchain pointer at SP.  Drawback: even calls that do not require stack arguments now require extra stack pointer manipulation.
4. Do not maintain the initial backchain pointer on the stack at all.  Instead, keep initial backchain pointer in some newly reserved register.  Drawback: wastes a register, no longer compatible with external unwinding tools.

Unless I'll get any better ideas, right now 2) seems like the best trade-off to me.  But I'll continue to think about this ...

**Comment by fitzgen** [comment]
@uweigand thanks so much for the feedback! Still largely chewing on and processing it. One immediate thought below.

> [ B.t.w. callee-pops seems fundamentally incompatible with variable argument lists, so I assume this ABI makes use of the fact that those are not supported in wasmtime, right? ]

I haven't thought about variable argument lists at all because they both aren't supported in Cranelift right now and are not required for implementing any Wasm feature either.

That said, I'm not sure I follow why callee-pops is fundamentally incompatible with variable argument lists. It seems like we could extend this calling convention and pass around a dynamic argument count and callees could use that to clean up stack arguments. But maybe I am missing some flaw in that approach?

Either way, not super relevant since I don't expect Cranelift to support varargs anytime soon, and we can always conditionally support it for some calling conventions and not others.

**Comment by uweigand** [comment]
> I haven't thought about variable argument lists at all because they both aren't supported in Cranelift right now and are not required for implementing any Wasm feature either.
> 
> That said, I'm not sure I follow why callee-pops is fundamentally incompatible with variable argument lists. It seems like we could extend this calling convention and pass around a dynamic argument count and callees could use that to clean up stack arguments. But maybe I am missing some flaw in that approach?

Hmm, I was thinking about the default system ABI requirements, where you usually not only need to support varargs, but need to support varargs using the *same* calling convention as any non-varargs call.  That is needed to support pre-ANSI C code, where it is OK to call a function without prototype, so the compiler does not even know whether the target is vararg or not - as long as the actual arguments passed are correct, it has to work either way.

But I guess that really isn't a requirement for a new wasmtime calling convention; even if there ever was a need to support varargs in wasmtime, I'm pretty sure that specific requirement will not apply, so we can indeed choose a different convention for varargs (e.g. with an extra count).

**Comment by fitzgen** [comment]
@uweigand thanks again for your feedback! I've finally had a chance to integrate some of your points into the document. Mostly I clarified that this is not intended to be prescriptive or say that all architectures must match exactly what x86-64 does. We just worked out most of the details there because on-stack return addresses pose a particular problem with placement of the stack arguments and because it is what I am personally most familiar with. I additionally enumerated the most important properties we need to preserve, as you suggested: callee pops and fast stack walking.

**Comment by fitzgen** [comment]
# Motion to Finalize

**Disposition: Merge**

It seems like we have mostly come to consensus here and the bits that we haven't hammered out 100% are things we can do during the implementation phase. As such, I am moving that we merge this RFC.

------------------

Note: I don't want to trawl through commit logs and try and figure out which committer is part of which BA organization, so I'm not making a stakeholder check list. Just approve the PR  or comment to sign off, or alternatively leave a comment detailing your objections if you have them. I'll manually keep track of whether we have multi-org sign off and all that as the comments come in.

As always, details on the RFC process can be found here: https://github.com/bytecodealliance/rfcs/blob/main/accepted/rfc-process.md#making-a-decision-merge-or-close

**Review by alexcrichton — APPROVED** [review]

**Review by bjorn3 — APPROVED** [review]
(I'm unaffiliated)

**Review by cfallin — APPROVED** [review]

**Review by uweigand — APPROVED** [review]
Thanks for those changes.  LGTM!

**Review by tschneidereit — APPROVED** [review]

**Comment by fitzgen** [comment]
As there have been signoffs from at least two people representing different stakeholder organizations, this RFC is entering its 10 day

# Final Comment Period

and the last day to raise objections before this can merge is 2023-02-14.

**Comment by akirilov-arm** [comment]
> I've added an open question about the interaction this might have with hardware CFI features (specifically Intel CET and its shadow stack, I don't think this is incompatible with ARM's pointer authentication, AFAIK).

I haven't digested the full proposal, but I would like to comment on the point above - when pointer authentication is used, return addresses are signed using their location on the stack as a modifier (what is referred to as having _address diversity_). If I am not mistaken, an implementation of the proposal might move the return address to a different location on the stack in certain situations. This means that it would need to be authenticated with one of the `AUTI*` instructions (to verify that it hasn't been tampered with) and then immediately resigned before being stored to memory.

**Comment by fitzgen** [comment]
Thanks for the clarification @akirilov-arm!

> This means that it would need to be authenticated with one of the `AUTI*` instructions (to verify that it hasn't been tampered with) and then immediately resigned before being stored to memory.

Good to know that the RA is movable so long as we make sure to re-authenticate and re-sign.

**Comment by akirilov-arm** [comment]
There is one more CFI detail on AArch64, but this time it concerns the interaction between BTI and tail calls - if the latter are done via an indirect branch, i.e. `BR`, then the register parameter must be either `X16` or `X17`, otherwise there will be a processor exception when the callee starts executing. As an example, currently [the trampolines](https://github.com/bytecodealliance/wasmtime/blob/main/crates/runtime/src/trampolines/aarch64.rs#L30) are written with this constraint in mind.

**Comment by Robbepop** [comment]
Sorry to be late to the show here but I know `wasmi` is already included in the Wasmtime differential fuzzing and quite recently `wasmi` got support for `tail-calls` so we could find bugs in either implementation by upgrading the `wasmi` fuzzing in Wasmtime maybe. :) Note that `wasmi` is much simpler than Wasmtime since it is just an interpreter.

**Comment by fitzgen** [comment]
> As there have been signoffs from at least two people representing different stakeholder organizations, this RFC is entering its 10 day
> # Final Comment Period
> 
> and the last day to raise objections before this can merge is 2023-02-24.

The FCP completed while I was on vacation, and no concerns have been raised, so I will now belatedly merge this PR!

Thanks for the discussion and input everybody!

**Pull request merged**
